src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -121,26 +143,49 @@ private IModelBinder CreateBinderCore(DefaultModelBinderProviderContext provider
                 }
             }
 
-            if (result == null && stack.Count > 1)
+            // If the PlaceholderBinder was created, then it means we recursed. Hook it up to the 'real' binder.
+            var placeholderBinder = collection[key] as PlaceholderBinder;
+            if (placeholderBinder != null)
             {
-                // Use a no-op binder if we're below the top level. At the top level, we throw.
-                result = NoOpBinder.Instance;
+                // It's also possible that user code called into `CreateBinder` but then returned null, we don't
+                // want to create something that will null-ref later so just hook this up to the no-op binder.
```

Comment:
Curious: Can't the `NoOpBinder` case be hit whenever a type w/o a binder is referenced recursively?